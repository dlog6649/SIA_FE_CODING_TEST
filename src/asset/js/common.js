const tagNm = {
    LINE: 'line'
    ,CIRCLE: 'circle'
    ,RECT: 'rect'
    ,TEXT: 'text'
    ,FOREIGNOBJECT: 'foreignObject'
    ,IMAGE: 'image'
    ,G: 'g'
    ,SPAN: 'SPAN'
}


export const parseTransform = obj => {
    if (obj.tagName === tagNm.IMAGE) {
      let transform = obj.getAttribute('transform').split(' ');
      let x = parseFloat(transform[0].substring(10));
      let y = parseFloat(transform[1].split(')')[0]);
      let scale = parseFloat(transform[2].substring(6).split(')')[0]);
      return {x: x, y: y, scale: scale}
    }
    else if (obj.tagName === tagNm.G) {
      let transform = obj.getAttribute('transform').split(' ');
      let x = parseFloat(transform[0].substring(10));
      let y = parseFloat(transform[1].split(')')[0]);
      let scale = parseFloat(transform[2].substring(6).split(')')[0]);
      let deg = parseInt(transform[3].substring(7));
      let rotX = parseFloat(transform[4]);
      let rotY = parseFloat(transform[5].split(')')[0]);
      let w = parseFloat(obj.firstChild.getAttribute('width'));
      let h = parseFloat(obj.firstChild.getAttribute('height'));
      return {x: x, y: y, deg: deg, rotX: rotX, rotY: rotY, w: w, h: h, scale: scale};
    }
}