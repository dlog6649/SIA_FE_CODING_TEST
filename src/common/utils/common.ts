const tagNm = {
  LINE: "line",
  CIRCLE: "circle",
  RECT: "rect",
  TEXT: "text",
  FOREIGNOBJECT: "foreignObject",
  IMAGE: "image",
  G: "g",
  SPAN: "SPAN",
}

export const parseTransform = (obj: any) => {
  let transform = ""
  let x = 0
  let y = 0
  let scale = 0
  let deg = 0
  let rotX = 0
  let rotY = 0
  let w = 0
  let h = 0
  if (obj.tagName === tagNm.IMAGE) {
    transform = obj.getAttribute("transform").split(" ")
    x = parseFloat(transform[0].substring(10))
    y = parseFloat(transform[1].split(")")[0])
    scale = parseFloat(transform[2].substring(6).split(")")[0])
  } else if (obj.tagName === tagNm.G) {
    transform = obj.getAttribute("transform").split(" ")
    x = parseFloat(transform[0].substring(10))
    y = parseFloat(transform[1].split(")")[0])
    scale = parseFloat(transform[2].substring(6).split(")")[0])
    deg = parseInt(transform[3].substring(7))
    rotX = parseFloat(transform[4])
    rotY = parseFloat(transform[5].split(")")[0])
    w = parseFloat(obj.firstChild.getAttribute("width"))
    h = parseFloat(obj.firstChild.getAttribute("height"))
  }
  return { x, y, deg, rotX, rotY, w, h, scale }
}

export const parseTransformG = (transform: string) => {
  console.log(transform)
  const attributeList = transform.split(" ")
  console.log(attributeList)
  // const x = parseFloat(attributeList[0].substring(10))
  // const y = parseFloat(attributeList[1].split(")")[0])
  // const scale = parseFloat(attributeList[2].substring(6).split(")")[0])
  // const deg = parseInt(attributeList[3].substring(7))
  // const rotX = parseFloat(attributeList[4])
  // const rotY = parseFloat(attributeList[5].split(")")[0])
  // return { x, y, deg, rotX, rotY }
  return { x: 123, y: 123 }
}

export function throttle(func: any, wait: number, options: any) {
  let context: any
  let args: IArguments | null
  let result: any
  let timeout: NodeJS.Timeout | null = null
  let previous = 0
  if (!options) {
    options = {}
  }
  const later = function () {
    previous = options.leading === false ? 0 : Date.now()
    timeout = null
    result = func.apply(context, args)
    if (!timeout) {
      context = args = null
    }
  }

  return function (this: any) {
    const now = Date.now()
    if (!previous && options.leading === false) {
      previous = now
    }
    const remaining = wait - (now - previous)
    context = this // if delete 'this' parameter then error occur
    args = arguments
    if (remaining <= 0 || remaining > wait) {
      if (timeout) {
        clearTimeout(timeout)
        timeout = null
      }
      previous = now
      result = func.apply(context, args)
      if (!timeout) {
        context = args = null
      }
    } else if (!timeout && options.trailing !== false) {
      timeout = setTimeout(later, remaining)
    }
    return result
  }
}

export const pauseEvent = (evt: any) => {
  if (evt.stopPropagation) {
    evt.stopPropagation()
  }
  if (evt.preventDefault) {
    evt.preventDefault()
  }
  evt.cancelBubble = true
  evt.returnValue = false
  return false
}

export const getLabelState = (label: SVGGElement) => {
  const id = Number(label.dataset.id)
  const name = label.dataset.name
  const tf = parseTransform(label)
  const data = { x: tf.x, y: tf.y, w: tf.w, h: tf.h, deg: tf.deg }

  const theta = (Math.PI / 180) * tf.deg
  const cos_t = Math.cos(theta)
  const sin_t = Math.sin(theta)

  const c_x = tf.x + tf.rotX
  const c_y = tf.y + tf.rotY

  const nw_x = tf.x
  const nw_y = tf.y
  const ne_x = tf.x + tf.w
  const ne_y = tf.y
  const se_x = tf.x + tf.w
  const se_y = tf.y + tf.h
  const sw_x = tf.x
  const sw_y = tf.y + tf.h

  let nw_xp = (nw_x - c_x) * cos_t - (nw_y - c_y) * sin_t + c_x
  let nw_yp = (nw_x - c_x) * sin_t + (nw_y - c_y) * cos_t + c_y
  let ne_xp = (ne_x - c_x) * cos_t - (ne_y - c_y) * sin_t + c_x
  let ne_yp = (ne_x - c_x) * sin_t + (ne_y - c_y) * cos_t + c_y
  let se_xp = (se_x - c_x) * cos_t - (se_y - c_y) * sin_t + c_x
  let se_yp = (se_x - c_x) * sin_t + (se_y - c_y) * cos_t + c_y
  let sw_xp = (sw_x - c_x) * cos_t - (sw_y - c_y) * sin_t + c_x
  let sw_yp = (sw_x - c_x) * sin_t + (sw_y - c_y) * cos_t + c_y

  nw_xp = parseFloat(nw_xp.toFixed(2))
  nw_yp = parseFloat(nw_yp.toFixed(2))
  ne_xp = parseFloat(ne_xp.toFixed(2))
  ne_yp = parseFloat(ne_yp.toFixed(2))
  se_xp = parseFloat(se_xp.toFixed(2))
  se_yp = parseFloat(se_yp.toFixed(2))
  sw_xp = parseFloat(sw_xp.toFixed(2))
  sw_yp = parseFloat(sw_yp.toFixed(2))

  // coordinates
  // 0 1
  // 3 2
  const coordinates = []
  coordinates.push({ x: nw_xp, y: nw_yp })
  coordinates.push({ x: ne_xp, y: ne_yp })
  coordinates.push({ x: se_xp, y: se_yp })
  coordinates.push({ x: sw_xp, y: sw_yp })

  return { id: id, name: name, coordinates: coordinates, data: data }
}

export const generateUUID = () => {
  let dt = new Date().getTime()
  const uuid = "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, (c) => {
    const r = (dt + Math.random() * 16) % 16 | 0
    dt = Math.floor(dt / 16)
    return (c == "x" ? r : (r & 0x3) | 0x8).toString(16)
  })
  return uuid
}
