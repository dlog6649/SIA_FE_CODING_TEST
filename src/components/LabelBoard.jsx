import React, { useState, useEffect } from 'react';
import LabelCtxMenu from './LabelCtxMenu';
import * as LabelMain from '../labeling-tool/LabelMain';
import { redrawImage, redrawLabels } from '../labeling-tool/LabelCreator';
import { compareImage, compareLabels, compareIds } from '../labeling-tool/LabelCompare';

export var _props;
export var _setScale;


export default function LabelBoard(props) {
    const [scale, setScale] = useState(1);

    useEffect(() => {
        console.log('LabelBoard useEffect: []');

        _props = props;
        _setScale = setScale;

        LabelMain.initialize();

        return () => LabelMain.finalize();
    }, []);


    useEffect(() => {
        console.log('LabelBoard useEffect: [props.mode]: ', props.mode);

        if (LabelMain.getMode() === props.mode) {
            console.log('useEffect: [props.mode]: returned ');
            return;
        }

        LabelMain.setMode(props.mode);

    }, [props.mode]);


    useEffect(() => {
        console.log('LabelBoard useEffect: [props.image]: ');

        if (compareImage(props.image)) {
            console.log('LabelBoard useEffect: [props.image]: returned');
            return;
        }

        redrawImage(props.currentImgURL, props.image);

    }, [props.image]);


    useEffect(() => {
        console.log('LabelBoard useEffect: [props.labels]: ');

        if (compareLabels(props.labels)) {
            console.log('LabelBoard useEffect: [props.labels]: returned');
            return;
        }

        redrawLabels(props.labels);

    }, [props.labels]);


    useEffect(() => {
        console.log('LabelBoard useEffect: [props.selectedLabelsIds]: ');

        if (compareIds(props.selectedLabelsIds)) {
            console.log('LabelBoard useEffect: [props.selectedLabelsIds]: returned');
            return;
        }

        LabelMain.createAnchorsInSelectedLabelsIds(props.selectedLabelsIds);

    }, [props.selectedLabelsIds]);


    const changeScale = e => {
        setScale(parseFloat(e.target.value));
    }


    const changeSliderCSS = e => {
        let val = e.target.value;
        if (val <= 0.1) {
            val = 0;
        }
        e.target.style.background = 'linear-gradient(to right, #333333 0%, #333333 ' + val * 50 + '%, #dedede ' + val * 50 + '%, #dedede 100%)';
    }


    return (
        <div className="label-board">
            <svg id="svg" width="100%" height="100%" data-testid="testSvg">
                <defs>
                    <filter id="f1">
                        <feDropShadow dx="-1" dy="1" stdDeviation="2.5" floodColor="gray" />
                    </filter>
                </defs>
            </svg>
            <div className="label-board-scaler">
                <img className="scaler-plus btn-img" src={require('../asset/images/plus.png')} alt="+" />
                <input className="scaler-range" type="range" data-testid="testScaler" orient="vertical" value={scale} onInput={changeSliderCSS} onChange={changeScale} min="0.1" max="2" step="0.1" />
                <img className="scaler-minus btn-img" src={require('../asset/images/minus.png')} alt="-" />
            </div>
            <LabelCtxMenu />
        </div>
    );
}
