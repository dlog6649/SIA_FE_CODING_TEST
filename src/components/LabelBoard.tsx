import React, { useState, useEffect } from 'react';
import LabelCtxMenu from './LabelCtxMenu';
import * as LabelMain from '../labeling-tool/LabelMain';
import { redrawImage, redrawLabels } from '../labeling-tool/LabelCreator';
import { compareImage, compareLabels, compareIds } from '../labeling-tool/LabelCompare';

export var _props: Props;
export var _setScale: React.Dispatch<React.SetStateAction<number>>;

interface Label {
    id: number
    name: string
    coordinates: Array<{x:number, y: number}>
    data: {x: number, y: number, w: number, h: number, deg: number}
}

interface Props {
    mode: string
    image: any
    currentImgURL: string
    labels: Array<Label>
    selectedLabelsIds: number[]
}

export default function LabelBoard(props: Props) {
    const [scale, setScale] = useState(1);

    useEffect(() => {
        console.log('LabelBoard useEffect: []');
        _props = props;
        _setScale = setScale;
        LabelMain.initialize();
        return () => LabelMain.finalize();
    }, []);

    useEffect(() => {
        console.log('LabelBoard useEffect: [props.mode]: ', props.mode);
        if (LabelMain.getMode() === props.mode) {
            console.log('useEffect: [props.mode]: returned ');
            return;
        }
        LabelMain.setMode(props.mode);
    }, [props.mode]);

    useEffect(() => {
        console.log('LabelBoard useEffect: [props.image]: ');
        if (compareImage(props.image)) {
            console.log('LabelBoard useEffect: [props.image]: returned');
            return;
        }
        redrawImage(props.currentImgURL, props.image);
    }, [props.image]);

    useEffect(() => {
        console.log('LabelBoard useEffect: [props.labels]: ');
        if (compareLabels(props.labels)) {
            console.log('LabelBoard useEffect: [props.labels]: returned');
            return;
        }
        redrawLabels(props.labels);
    }, [props.labels]);

    useEffect(() => {
        console.log('LabelBoard useEffect: [props.selectedLabelsIds]: ');
        if (compareIds(props.selectedLabelsIds)) {
            console.log('LabelBoard useEffect: [props.selectedLabelsIds]: returned');
            return;
        }
        LabelMain.createAnchorsInSelectedLabelsIds(props.selectedLabelsIds);
    }, [props.selectedLabelsIds]);

    const changeScale = (evt: any): void => {
        setScale(parseFloat(evt.target.value));
    }

    const changeSliderCSS = (evt: any): void => {
        let value = evt.target.value;
        if (value <= 0.1) {
            value = 0;
        }
        evt.target.style.background = 'linear-gradient(to right, #333333 0%, #333333 ' + value * 50 + '%, #dedede ' + value * 50 + '%, #dedede 100%)';
    }

    return (
        <div className="label-board">
            <svg id="svg" width="100%" height="100%" data-testid="testSvg">
                <defs>
                    <filter id="f1">
                        <feDropShadow dx="-1" dy="1" stdDeviation="2.5" floodColor="gray" />
                    </filter>
                </defs>
            </svg>
            <div className="label-board-scaler">
                <img className="scaler-plus btn-img" src={require('../asset/images/plus.png')} alt="+" />
                <input className="scaler-range" orient="vertical" type="range" data-testid="testScaler" value={scale} onInput={changeSliderCSS} onChange={changeScale} min="0.1" max="2" step="0.1" />
                <img className="scaler-minus btn-img" src={require('../asset/images/minus.png')} alt="-" />
            </div>
            <LabelCtxMenu />
        </div>
    );
}
