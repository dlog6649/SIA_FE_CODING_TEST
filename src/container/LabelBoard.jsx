import LabelBoard from '../components/LabelBoard';
import { connect } from 'react-redux';
import { selectLabels, createLabels, updateLabels, updateImgLabels, deleteLabels } from '../modules/annotator';


const mapStateToProps = state => {
    return {
        mode: state.annotator.mode
        ,currentImgURL: state.annotator.currentImgURL
        ,selectedLabelsIds: state.annotator.selectedLabelsIds
        ,image: state.annotator.images[state.annotator.currentImgURL]
        ,labels: state.annotator.labels[state.annotator.currentImgURL]
    };
}

const mapDispatchToProps = dispatch => {
    return {
        selectLabels: selectedLabelsIds => dispatch(selectLabels(selectedLabelsIds))
        ,createLabels: labels => dispatch(createLabels(labels))
        ,updateLabels: (labels, selectedLabelsIds) => dispatch(updateLabels(labels, selectedLabelsIds))
        ,updateImgLabels: (image, labels, selectedLabelsIds) => dispatch(updateImgLabels(image, labels, selectedLabelsIds))
        ,deleteLabels: selectedLabelsIds => dispatch(deleteLabels(selectedLabelsIds))
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(LabelBoard);