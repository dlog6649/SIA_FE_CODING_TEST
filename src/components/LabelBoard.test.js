import React from 'react';
import { render, fireEvent } from "@testing-library/react";
import LabelBoard from './LabelBoard';
import { LABEL_SELECT_MODE, LABEL_CREATE_MODE } from '../modules/annotator';
import { parseTransform } from '../asset/js/common';


describe('LabelBoard 라벨링툴 테스트 시작', () => {
  let wrapper = null;
  const _currentImgURL = 'http://sample.png';
  const _selectedLabelsIds = [];
  const _image = {title: 'Lorem ipsum', x: 0, y: 0, scale: 1};
  const _labels = [
    {id: 0, name: 'sea', coordinates: [{x: 0, y: 0}, {x: 50, y: 0}, {x: 50, y: 50}, {x: 0, y: 50}], data: {x: 0, y: 0, w: 50, h: 50, deg: 0}}
    ,{id: 1, name: 'woods', coordinates: [{x: 50, y: 0}, {x: 100, y: 0}, {x: 100, y: 50}, {x: 50, y: 50}], data: {x: 50, y: 0, w: 50, h: 50, deg: 0}}
    ,{id: 2, name: 'building', coordinates: [{x: 0, y: 50}, {x: 50, y: 50}, {x: 50, y: 100}, {x: 0, y: 50}], data: {x: 0, y: 50, w: 50, h: 50, deg: 0}}
  ];
  const mouseDown = new Event('mousedown');
  Object.defineProperty(mouseDown, 'button', {get: () => 0});
  Object.defineProperty(mouseDown, 'offsetX', {get: () => 0});
  Object.defineProperty(mouseDown, 'offsetY', {get: () => 0});
  const mouseMove = new Event('mousemove');
  Object.defineProperty(mouseMove, 'offsetX', {get: () => 200});
  Object.defineProperty(mouseMove, 'offsetY', {get: () => 200});

  it('렌더링이 된 결과가 기존 스냅샷과 비교해서 일치해야 한다.', () => {
    wrapper = render(
      <LabelBoard 
        mode={LABEL_SELECT_MODE} currentImgURL={_currentImgURL} selectedLabelsIds={_selectedLabelsIds} image={_image} labels={_labels}
      />
    );
    expect(wrapper.baseElement).toMatchSnapshot();
  });

  describe('라벨 생성모드', () => {
    it('라벨링보드에 드래그를 하면 라벨과 클래스입력창이 생성되고, 생성된 라벨은 디스패치된다.', () => {
      const mockCreateLabels = jest.fn();
      wrapper = render(
        <LabelBoard 
          mode={LABEL_CREATE_MODE} currentImgURL={_currentImgURL} selectedLabelsIds={[]} image={_image} labels={[]} 
          createLabels={mockCreateLabels}
        />
      );
      let labelingBoard = wrapper.getByTestId('testSvg');
  
      // 라벨링보드에 드래그
      fireEvent(labelingBoard, mouseDown);
      fireEvent(labelingBoard, mouseMove);
      fireEvent.mouseUp(labelingBoard);
      
      let testLabels = wrapper.getAllByTestId('testLabel');
      let inputForms = wrapper.getAllByPlaceholderText('Input Class name');
      let createLabelsEventByMouseUp = mockCreateLabels.mock.calls;
      let [labels] = createLabelsEventByMouseUp[0];
      expect(testLabels).toHaveLength(1);
      expect(inputForms).toHaveLength(1);
      expect(labels).toHaveLength(1);
    });

    it('생성된 라벨의 클래스 입력창에 문자를 입력 후 엔터키를 누르면 클래스명이 부여되고, 해당 라벨은 디스패치된다.', () => {
      const mockCreateLabels = jest.fn();
      const mockUpdateLabels = jest.fn();
      wrapper = render(
        <LabelBoard 
          mode={LABEL_CREATE_MODE} currentImgURL={_currentImgURL} selectedLabelsIds={[]} image={_image} labels={[]} 
          createLabels={mockCreateLabels} updateLabels={mockUpdateLabels}
        />
      );
      const labelingBoard = wrapper.getByTestId('testSvg');
  
      fireEvent(labelingBoard, mouseDown);
      fireEvent(labelingBoard, mouseMove);
      fireEvent.mouseUp(labelingBoard);
      
      let inputForm = wrapper.getByPlaceholderText('Input Class name');
      inputForm.value = 'new class name';
      fireEvent.keyDown(inputForm, {keyCode: 13}); // enter키 입력
      
      let updateLabelsEventByKeyDown = mockUpdateLabels.mock.calls[0];
      let [labels] = updateLabelsEventByKeyDown;
      expect(labels[0].dataset.name).toEqual('new class name');
    });
  });

  describe('라벨 선택모드', () => {
    it('라벨을 클릭하면 라벨이 선택되며 앵커가 표시된다. 그리고 선택된 라벨의 ID는 디스패치된다.', () => {
      const mockUpdateLabels = jest.fn();

      // 기본값으로 라벨 3개 전달
      wrapper = render(
        <LabelBoard 
          mode={LABEL_SELECT_MODE} currentImgURL={_currentImgURL} selectedLabelsIds={[]} image={_image} labels={_labels} 
          updateLabels={mockUpdateLabels}
        />
      );
      let testLabels = wrapper.getAllByTestId('testLabel');
  
      // 0번 라벨 클릭
      fireEvent.mouseDown(testLabels[0].firstChild);
      fireEvent.mouseUp(testLabels[0].firstChild);

      let updateLabelsEventByMouseUp = mockUpdateLabels.mock.calls;
      let [labels, selectedLabelsIds] = updateLabelsEventByMouseUp[0];
      expect(testLabels[0].classList.contains('selected')).toBe(true);
      expect(selectedLabelsIds).toEqual([0]);
    });
  
    it('컨트롤키를 누르고 라벨 클릭 시 다중선택이 가능해진다. 그리고 선택된 라벨들의 ID는 디스패치된다.', () => {
      const mockUpdateLabels = jest.fn();

      // 기본값으로 라벨 3개 전달
      wrapper = render(
        <LabelBoard 
          mode={LABEL_SELECT_MODE} currentImgURL={_currentImgURL} selectedLabelsIds={[]} image={_image} labels={_labels} 
          updateLabels={mockUpdateLabels}
        />
      );
      let testLabels = wrapper.getAllByTestId('testLabel');
  
      // 0, 1번 라벨 클릭
      fireEvent.mouseDown(testLabels[0].firstChild, {ctrlKey: true});
      fireEvent.mouseUp(testLabels[0].firstChild);
      fireEvent.mouseDown(testLabels[1].firstChild, {ctrlKey: true});
      fireEvent.mouseUp(testLabels[1].firstChild);
  
      let updateLabelsEventByMouseUp = mockUpdateLabels.mock.calls;
      let [labels, selectedLabelsIds] = updateLabelsEventByMouseUp[1];
      expect(testLabels[0].classList.contains('selected')).toBe(true);
      expect(testLabels[1].classList.contains('selected')).toBe(true);
      expect(selectedLabelsIds).toEqual([0, 1]);
    });
  
    it('라벨을 클릭 후 delete키 또는 backspace키를 누를 시 삭제된다. 그리고 삭제된 라벨들의 ID는 디스패치된다.', () => {
      const mockUpdateLabels = jest.fn();
      const mockDeleteLabels = jest.fn();

      // 기본값으로 라벨 3개 전달
      wrapper = render(
        <LabelBoard 
          mode={LABEL_SELECT_MODE} currentImgURL={_currentImgURL} selectedLabelsIds={[]} image={_image} labels={_labels} 
          updateLabels={mockUpdateLabels} deleteLabels={mockDeleteLabels}
        />
      );
      let testLabels = wrapper.getAllByTestId('testLabel');
  
      // 0번 라벨 클릭 후 delete키 입력
      fireEvent.mouseDown(testLabels[0].firstChild);
      fireEvent.mouseUp(testLabels[0].firstChild);
      fireEvent.keyDown(document, {which: 46, keyCode: 46});
  
      // 1번 라벨 클릭 후 backspace키 입력
      fireEvent.mouseDown(testLabels[1].firstChild);
      fireEvent.mouseUp(testLabels[1].firstChild);
      fireEvent.keyDown(document, {which: 8, keyCode: 8});
  
      let deleteLabelsEventByMouseDown = mockDeleteLabels.mock.calls;
      let [selectedLabelsIds] = deleteLabelsEventByMouseDown[0];
      expect(selectedLabelsIds).toEqual([0]);
  
      [selectedLabelsIds] = deleteLabelsEventByMouseDown[1];
      expect(selectedLabelsIds).toEqual([1]);
    });
  
    it('라벨을 드래그시 선택된 라벨의 위치가 이동된다. 그리고 라벨들은 디스패치된다.', () => {
      const mockUpdateLabels = jest.fn();

      // 기본값으로 라벨 3개 전달
      wrapper = render(
        <LabelBoard 
          mode={LABEL_SELECT_MODE} currentImgURL={_currentImgURL} selectedLabelsIds={[]} image={_image} labels={_labels} 
          updateLabels={mockUpdateLabels}
        />
      );
      let labelingBoard = wrapper.getByTestId('testSvg');
      let testLabels = wrapper.getAllByTestId('testLabel');
      let beforeLabel = parseTransform(testLabels[0]);
      expect(beforeLabel.x).toBe(0);
      expect(beforeLabel.y).toBe(0);
  
      // 0번 라벨 드래그
      fireEvent(testLabels[0].firstChild, mouseDown);
      fireEvent(labelingBoard, mouseMove);
      fireEvent.mouseUp(testLabels[0].firstChild);
  
      let updateLabelsEventByMouseDown = mockUpdateLabels.mock.calls;
      let [labels, selectedLabelsIds] = updateLabelsEventByMouseDown[0];
      let afterLabel = parseTransform(labels[0]);
      expect(afterLabel.x).toBe(200);
      expect(afterLabel.y).toBe(200);
    });
  
    it('라벨을 클릭하여 표시된 앵커를 드래그시 크기가 변경된다. 그리고 라벨들은 디스패치된다.', () => {
      const mockUpdateLabels = jest.fn();

      // 기본값으로 라벨 3개 전달
      wrapper = render(
        <LabelBoard 
          mode={LABEL_SELECT_MODE} currentImgURL={_currentImgURL} selectedLabelsIds={[]} image={_image} labels={_labels} 
          updateLabels={mockUpdateLabels}
        />
      );
      let labelingBoard = wrapper.getByTestId('testSvg');
      let testLabels = wrapper.getAllByTestId('testLabel');
      // 0번 라벨의 위치 (0, 0)
      let beforeWidth = parseFloat(testLabels[0].firstChild.getAttribute('width'));
      let beforeHeight = parseFloat(testLabels[0].firstChild.getAttribute('height'));
      expect(beforeWidth).toBe(50);
      expect(beforeHeight).toBe(50);
  
      // 0번 라벨 클릭
      fireEvent(testLabels[0].firstChild, mouseDown);
      fireEvent.mouseUp(testLabels[0].firstChild);
  
      // 좌측상단 앵커를 (200, 200)으로 드래그
      let nwAnchor = testLabels[0].firstChild.nextSibling.nextSibling.nextSibling;
      fireEvent(nwAnchor, mouseDown);
      fireEvent(labelingBoard, mouseMove);
      fireEvent.mouseUp(nwAnchor);
  
      let updateLabelsEventByMouseDown = mockUpdateLabels.mock.calls;
      let [labels, selectedLabelsIds] = updateLabelsEventByMouseDown[1];
      let afterWidth = parseFloat(labels[0].firstChild.getAttribute('width'));
      let afterHeight = parseFloat(labels[0].firstChild.getAttribute('height'));
      expect(afterWidth).toBe(150);
      expect(afterHeight).toBe(150);
    });
  });

  describe('아무 모드에서나 가능한 기능 테스트', () => {
    it('우측하단의 Slider나 마우스휠을 이용해서 이미지와 라벨들을 확대 및 축소할 수 있다. 그리고 이미지와 라벨들은 디스패치된다.', () => {
      const mockCreateLabels = jest.fn();
      const mockUpdateLabels = jest.fn();
      const mockUpdateImgLabels = jest.fn();
      const mockSelectLabels = jest.fn();
  
      // 기본값으로 라벨 3개 전달
      wrapper = render(
        <LabelBoard 
          mode={LABEL_SELECT_MODE} currentImgURL={_currentImgURL} selectedLabelsIds={[]} image={_image} labels={_labels} 
          createLabels={mockCreateLabels} updateLabels={mockUpdateLabels} selectLabels={mockSelectLabels} updateImgLabels={mockUpdateImgLabels} 
        />
      );  
      let labelingBoard = wrapper.getByTestId('testSvg');
      let testScaler = wrapper.getByTestId('testScaler');
      let testImg = wrapper.getByTestId('testImg');
      let testLabels = wrapper.getAllByTestId('testLabel');
  
      // 슬라이더의 scale값을 1.5로 변경
      testScaler.value = 1.5;
      fireEvent.change(testScaler);
  
      let beforeTestImg = parseTransform(testImg);
      let beforeTestLabel = parseTransform(testLabels[0]);
      expect(beforeTestImg.scale).toBe(1.5);
      expect(beforeTestLabel.scale).toBe(1.5);
  
      // 라벨링보드에서 마우스휠 회전
      fireEvent(labelingBoard, new MouseEvent('mousewheel', {deltaY: -100}));
  
      let afterTestImg = parseTransform(testImg);
      let afterTestLabel = parseTransform(testLabels[0]);
      expect(afterTestImg.scale).toBe(1.6);
      expect(afterTestLabel.scale).toBe(1.6);
  
      let updateImgLabelsEventByMouseWheel = mockUpdateImgLabels.mock.calls;
      let [img, labels] = updateImgLabelsEventByMouseWheel[1];
      let afterImg = parseTransform(img);
      let afterLabel = parseTransform(labels[0]);
      expect(afterImg.scale).toBe(1.6);
      expect(afterLabel.scale).toBe(1.6);
    });
  
    it('스페이스바를 누른 상태에서 이미지를 드래그시 이미지와 라벨들의 위치가 이동된다. 그리고 이미지와 라벨들은 디스패치된다.', () => {
      const mockCreateLabels = jest.fn();
      const mockUpdateLabels = jest.fn();
      const mockUpdateImgLabels = jest.fn();
      const mockSelectLabels = jest.fn();

      // 기본값으로 라벨 3개 전달
      wrapper = render(
        <LabelBoard 
          mode={LABEL_SELECT_MODE} currentImgURL={_currentImgURL} selectedLabelsIds={[]} image={_image} labels={_labels} 
          createLabels={mockCreateLabels} updateLabels={mockUpdateLabels} selectLabels={mockSelectLabels} updateImgLabels={mockUpdateImgLabels} 
        />
      );
      let labelingBoard = wrapper.getByTestId('testSvg');
      let testImg = wrapper.getByTestId('testImg');
      let testLabels = wrapper.getAllByTestId('testLabel');
      let beforeImg = parseTransform(testImg);
      let beforeLabel = parseTransform(testLabels[0]);
      expect(beforeImg.x).toBe(0);
      expect(beforeImg.y).toBe(0);
      expect(beforeLabel.x).toBe(0);
      expect(beforeLabel.y).toBe(0);
  
      // 스페이스바를 누른 상태로 드래그
      fireEvent.keyDown(document, {which: 32, keyCode: 32});
      fireEvent(labelingBoard, mouseDown);
      fireEvent(labelingBoard, mouseMove);
      fireEvent.mouseUp(labelingBoard);
  
      let updateImgLabelsEventByMouseUp = mockUpdateImgLabels.mock.calls;
      let [img, labels] = updateImgLabelsEventByMouseUp[0];
      let afterImg = parseTransform(img);
      let afterLabel = parseTransform(labels[0]);
      expect(afterImg.x).toBe(200);
      expect(afterImg.y).toBe(200);
      expect(afterLabel.x).toBe(200);
      expect(afterLabel.y).toBe(200);
    });
  });
});
