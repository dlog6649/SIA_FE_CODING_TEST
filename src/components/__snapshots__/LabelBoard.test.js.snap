// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LabelBoard 라벨링툴 테스트 시작 렌더링이 된 결과가 기존 스냅샷과 비교해서 일치해야 한다. 1`] = `
<body>
  <div>
    <div
      class="label-board"
    >
      <svg
        data-testid="testSvg"
        height="100%"
        id="svg"
        width="100%"
      >
        <defs>
          <filter
            id="f1"
          >
            <fedropshadow
              dx="-1"
              dy="1"
              flood-color="gray"
              stdDeviation="2.5"
            />
          </filter>
        </defs>
        <image
          alt="sampleImg"
          data-testid="testImg"
          href="http://sample.png"
          id="img"
          transform="translate(0 0) scale(1)"
        />
        <g
          class="label"
          data-id="0"
          data-name="sea"
          data-testid="testLabel"
          transform="translate(0 0) scale(1) rotate(0 25 25)"
        >
          <rect
            cursor="move"
            fill="#5c6dda"
            fill-opacity="0.2"
            height="50"
            stroke="#5c6dda"
            stroke-width="3"
            width="50"
          />
        </g>
        <g
          class="label"
          data-id="1"
          data-name="woods"
          data-testid="testLabel"
          transform="translate(50 0) scale(1) rotate(0 25 25)"
        >
          <rect
            cursor="move"
            fill="#5c6dda"
            fill-opacity="0.2"
            height="50"
            stroke="#5c6dda"
            stroke-width="3"
            width="50"
          />
        </g>
        <g
          class="label"
          data-id="2"
          data-name="building"
          data-testid="testLabel"
          transform="translate(0 50) scale(1) rotate(0 25 25)"
        >
          <rect
            cursor="move"
            fill="#5c6dda"
            fill-opacity="0.2"
            height="50"
            stroke="#5c6dda"
            stroke-width="3"
            width="50"
          />
        </g>
      </svg>
      <div
        class="label-board-scaler"
      >
        <img
          alt="+"
          class="scaler-plus btn-img"
          src="plus.png"
        />
        <input
          class="scaler-range"
          data-testid="testScaler"
          max="2"
          min="0.1"
          orient="vertical"
          step="0.1"
          type="range"
          value="1"
        />
        <img
          alt="-"
          class="scaler-minus btn-img"
          src="minus.png"
        />
      </div>
      <div
        class="label-contextmenu"
      >
        <div
          class="item edit"
          id="edit"
        >
          <span
            class="item-name"
          >
            Edit Class
          </span>
          <span
            class="item-shortcut"
          >
            (TBD)
          </span>
        </div>
        <div
          class="item cut"
          id="cut"
        >
          <span
            class="item-name"
          >
            Cut
          </span>
          <span
            class="item-shortcut"
          >
            Ctrl + X
          </span>
        </div>
        <div
          class="item copy"
          id="copy"
        >
          <span
            class="item-name"
          >
            Copy
          </span>
          <span
            class="item-shortcut"
          >
            Ctrl + C
          </span>
        </div>
        <div
          class="item paste"
          id="paste"
        >
          <span
            class="item-name"
          >
            Paste
          </span>
          <span
            class="item-shortcut"
          >
            Ctrl + V
          </span>
        </div>
        <div
          class="item delete"
          id="delete"
        >
          <span
            class="item-name"
          >
            Delete
          </span>
          <span
            class="item-shortcut"
          >
            Del
          </span>
        </div>
      </div>
    </div>
  </div>
</body>
`;
